src/libimg/Image.cpp

```diff
@@ -275,6 +277,227 @@ void Image::drawImage(const Image& src, int dstX, int dstY) {
+        const float xgap = rfpart(x1 + 0.5);
+        xpx12 = int(xend);
+        const int ypx12 = ipart(yend);
+        if (steep) {
+            plot(ypx12,     xpx12, rfpart(yend) * xgap);
+            plot(ypx12 + 1, xpx12,  fpart(yend) * xgap);
+        } else {
+            plot(xpx12, ypx12,    rfpart(yend) * xgap);
+            plot(xpx12, ypx12 + 1, fpart(yend) * xgap);
+        }
+    }
+ 
+    if (steep) {
+        for (int x = xpx11 + 1; x < xpx12; x++) {
+            plot(ipart(intery),     x, rfpart(intery));
+            plot(ipart(intery) + 1, x,  fpart(intery));
+            intery += gradient;
+        }
+    } else {
+        for (int x = xpx11 + 1; x < xpx12; x++) {
+            plot(x, ipart(intery),     rfpart(intery));
+            plot(x, ipart(intery) + 1,  fpart(intery));
+            intery += gradient;
+        }
+    }
+}
+
+void Image::fillCircleCacheImage(int x_centre, int y_centre, int radius, uint32_t color) {
+    float d;
+    int alpha;
+    for (int y = y_centre - radius; y < y_centre + radius; y++) {
+        for (int x = x_centre - radius; x < x_centre + radius; x++) {
+            d = sqrt(pow((x - x_centre),2) + pow((y - y_centre),2));
+            if (d > (radius + 0.5))
+                alpha = 0; // Definitely outside
+            else if (d < (radius - 0.5))
+                alpha = 255; // Definitely inside
+            else
+                alpha = ((radius + 0.5) - d) * 255; // Border, so alpha blend for anti-aliasing
+            auto colorAA = (alpha << 24) | (color & 0x00FFFFFF);
+            blendPixel(x, y, colorAA);
+        }
+    }	
+}
+
+void Image::fillCircle(int x_centre, int y_centre, int radius, uint32_t color) {
+    // Due to compute complexity, all filled circles are rendered and cached with a 
+    // key consisting of the radius and color. So typically just a small image blend.
+    uint64_t key = (uint64_t)(radius) << 32 | color;
+    std::shared_ptr<img::Image> pImage;
+    auto it = circleCache.find(key);
+    if (it != circleCache.end()) {
+        pImage = it->second;
+    } else {
+        pImage = std::make_shared<img::Image>(radius*2+1, radius*2+1, 0x00FFFFFF);
+        pImage->fillCircleCacheImage(radius, radius, radius, color); 
+        circleCache.insert(std::make_pair(key, pImage));
+    }
+    blendImage0(*pImage, x_centre-radius, y_centre-radius);
+}
+
+// This class is a representation of a classic line equation ax + by + c
+// It is used in rectangle fills to determine whether a point is inside or outside
+// a rectangle bounding line. And how far away from the line it is so as to determine
+// appropriate alpha blend for anti-aliasing 
+class LineEquation {
+public:
+    LineEquation(int x1, int y1, int x2, int y2, int ref_x, int ref_y) {
+        a = y1 - y2;
+        b = x2 - x1;
+        c = x1*y2 - x2*y1;
+        normalAdjust = 10/sqrt(a*a + b*b);
+        refOffsetSign = (computeDistance(ref_x, ref_y) >= 0);
+    }
+    ~LineEquation(){

```

Comment:
In C++11, you can just write `~LineEquation() = default` to get this destructor, that's especially useful in header files so you don't need to put any code into the implementation file. But since the destructor is not virtual, you can just leave it out completely.